tests/virtctl/usbredir.go

```diff
@@ -0,0 +1,271 @@
+		})
+
+		It("Should fail when limit is reached", func() {
+			type session struct {
+				cancel context.CancelFunc
+				err    chan error
+			}
+			var tests []session
+			for i := range v1.UsbClientPassthroughMaxNumberOf + 1 {
+				Eventually(func(g Gomega) bool {
+					cmd := newVirtctlCommand("usbredir",
+						"--namespace", vmi.ObjectMeta.Namespace,
+						"--no-launch", vmi.ObjectMeta.Name)
+					connect := make(chan struct{})
+					ctx := context.WithValue(cmd.Context(), connectedKey, connect)
+					ctx, cancel := context.WithCancel(ctx)
+					cmd.SetContext(ctx)
+
+					err := make(chan error, 1)
+					go runConnectGoroutine(cmd, err)
+
+					if i == v1.UsbClientPassthroughMaxNumberOf {
+						// Last test is meant to fail.
+						tests = append(tests, session{cancel, err})
+						return true
+					}
+
+					// Till the last test, all sockets must be connected
+					select {
+					case <-connect:
+						tests = append(tests, session{cancel, err})
+						return true
+					case err := <-err:
+						Expect(err).To(MatchError(syscall.ECONNRESET))
+					}
+					cancel()
+					return false
+				}, 5*time.Second, 1*time.Second).Should(BeTrue())
+			}
+
+			numOfErrors := 0
+			for i := range v1.UsbClientPassthroughMaxNumberOf + 1 {
+				select {
+				case err := <-tests[i].err:
+					Expect(err).To(MatchError(ContainSubstring("websocket: bad handshake")))
+					numOfErrors++
+					tests[i].cancel()
+				case <-time.After(time.Second):
+					tests[i].cancel()
+				}
+			}
+			Expect(numOfErrors).To(Equal(1), "Only one connection should fail")
+		})
+
+		It("Should work several times", func() {
+			for range 4 * v1.UsbClientPassthroughMaxNumberOf {
+				Eventually(func(g Gomega) bool {
+					cmd := newVirtctlCommand("usbredir",
+						"--namespace", vmi.ObjectMeta.Namespace,
+						"--no-launch", vmi.ObjectMeta.Name)
+					connect := make(chan struct{})
+					ctx := context.WithValue(cmd.Context(), connectedKey, connect)
+					ctx, cancel := context.WithTimeout(ctx, time.Second)
+					cmd.SetContext(ctx)
+					defer cancel()
+
+					err := make(chan error, 1)
+					go runConnectGoroutine(cmd, err)
+
+					select {
+					case <-connect:
+						// Sent and Received message back. No errors.
+						return true
+					case err := <-err:
+						g.Expect(err).To(MatchError(syscall.ECONNRESET))
+					case <-ctx.Done():
+						g.Expect(ctx.Err()).To(MatchError(ContainSubstring("context canceled")))
+					}
+					return false
+				}, 5*time.Second, 1*time.Second).Should(BeTrue())
+			}
+		})
+	})
+})
+
+func runConnectGoroutine(cmd *cobra.Command, errch chan error) {
+	defer GinkgoRecover()
+	ctx := cmd.Context()
+	// To find ip/port to connect
+	rOut, wOut := io.Pipe()
+	cmd.SetOut(wOut)
+	defer rOut.Close()
+
+	// To find errors
+	rErr, wErr := io.Pipe()
+	cmd.SetErr(wErr)
+	defer rErr.Close()
+
+	// Make remote bufferred to not block select
+	remote := make(chan error, 1)
+	go func() {
+		defer GinkgoRecover()
+		defer close(remote)
```

Comment:
We can close `remote`, but not sure it is actually required.

Reply:
Removed